### ISSUE TITLE
Facing a C++ error when using a custom LM with CTC decoder

### ISSUE THREAD

**Issue by ziadloo** [pytorch/audio#3218]
### 🐛 Describe the bug

Following the tutorial on the [PyTorch website](https://pytorch.org/audio/stable/tutorials/asr_inference_with_ctc_decoder_tutorial.html), when I try to define a custom LM to present it to CTC decoder, I'll face an error which I believe is actually a C++ error that has bubbled up to python.

Consider the following code snippet:

```python
import torch
import torchaudio
from torchaudio.models.decoder import ctc_decoder
from torchaudio.utils import download_asset
from torchaudio.models.decoder import download_pretrained_files
from torchaudio.models.decoder import CTCDecoderLM, CTCDecoderLMState


bundle = torchaudio.pipelines.WAV2VEC2_ASR_BASE_10M
acoustic_model = bundle.get_model()

speech_file = download_asset("tutorial-assets/ctc-decoding/1688-142285-0007.wav")

waveform, sample_rate = torchaudio.load(speech_file)

if sample_rate != bundle.sample_rate:
    waveform = torchaudio.functional.resample(waveform, sample_rate, bundle.sample_rate)

files = download_pretrained_files("librispeech-4-gram")

actual_transcript = "i really was very much afraid of showing him how much shocked i was at some parts of what he said"
actual_transcript = actual_transcript.split()

emission, _ = acoustic_model(waveform)


class RandomLM(torch.nn.Module):
    def __init__(self):
        super().__init__()

    def forward(self, token_index: int) -> float:
        return torch.rand(1).item()


class CustomLM(CTCDecoderLM):
    """Create a Python wrapper around `language_model` to feed to the decoder."""
    def __init__(self, language_model: torch.nn.Module):
        CTCDecoderLM.__init__(self)
        self.language_model = language_model
        self.sil = -1  # index for silent token in the language model
        self.states = {}

        language_model.eval()

    def start(self, start_with_nothing: bool = False):
        state = CTCDecoderLMState()
        with torch.no_grad():
            score = self.language_model(self.sil)

        self.states[state] = score
        return state

    def score(self, state: CTCDecoderLMState, token_index: int):
        outstate = state.child(token_index)
        if outstate not in self.states:
            score = self.language_model(token_index)
            self.states[outstate] = score
        score = self.states[outstate]

        return outstate, score

    def finish(self, state: CTCDecoderLMState):
        return self.score(state, self.sil)

LM_WEIGHT = 3.23
WORD_SCORE = -0.26

dummy_decoder = ctc_decoder(
    lexicon=files.lexicon,
    tokens=files.tokens,
    lm=CustomLM(RandomLM()),
    nbest=3,
    beam_size=1500,
    lm_weight=LM_WEIGHT,
    word_score=WORD_SCORE,
)

random_result = dummy_decoder(emission)
dummy_transcript = " ".join(random_result[0][0].words).strip()

print(f"Transcript: {dummy_transcript}")
```

Most of the above code is coming from the mentioned tutorial. I just introduced the `RandomLM` class which returns a random number as the score whenever queried. This code faces the following error:

```
/home/mehran/.conda/envs/pytorch/lib/python3.10/site-packages/torchaudio/models/decoder/_ctc_decoder.py:62: UserWarning: The built-in flashlight integration is deprecated, and will be removed in future release. Please install flashlight-text. https://pypi.org/project/flashlight-text/ For the detail of CTC decoder migration, please see https://github.com/pytorch/audio/issues/3088.
  warnings.warn(

---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
Cell In[1], line 78
     66 WORD_SCORE = -0.26
     68 dummy_decoder = ctc_decoder(
     69     lexicon=files.lexicon,
     70     tokens=files.tokens,
   (...)
     75     word_score=WORD_SCORE,
     76 )
---> 78 random_result = dummy_decoder(emission)
     79 dummy_transcript = " ".join(random_result[0][0].words).strip()
     81 print(f"Transcript: {dummy_transcript}")

File ~/.conda/envs/pytorch/lib/python3.10/site-packages/torchaudio/models/decoder/_ctc_decoder.py:325, in CTCDecoder.__call__(self, emissions, lengths)
    323 for b in range(B):
    324     emissions_ptr = emissions.data_ptr() + float_bytes * b * emissions.stride(0)
--> 325     results = self.decoder.decode(emissions_ptr, lengths[b], N)
    327     nbest_results = results[: self.nbest]
    328     hypos.append(
    329         [
    330             CTCHypothesis(
   (...)
    337         ]
    338     )

RuntimeError: Tried to call pure virtual function "LM::start"
```

As Python does not have virtual functions, I assume this error actually belongs to the underlying C++ code. Anyways, the expected behaviour is to return a random string of characters.

### Versions

```
❯ python collect_env.py
Collecting environment information...
PyTorch version: 2.0.0
Is debug build: False
CUDA used to build PyTorch: 11.8
ROCM used to build PyTorch: N/A

OS: Manjaro Linux (x86_64)
GCC version: (GCC) 12.2.1 20230201
Clang version: 15.0.7
CMake version: version 3.25.2
Libc version: glibc-2.37

Python version: 3.10.10 (main, Mar 21 2023, 18:45:11) [GCC 11.2.0] (64-bit runtime)
Python platform: Linux-6.2.7-2-MANJARO-x86_64-with-glibc2.37
Is CUDA available: True
CUDA runtime version: 11.8.89
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: GPU 0: NVIDIA GeForce RTX 3090
Nvidia driver version: 525.89.02
cuDNN version: Probably one of the following:
/usr/lib/libcudnn.so.8.6.0
/usr/lib/libcudnn_adv_infer.so.8.6.0
/usr/lib/libcudnn_adv_train.so.8.6.0
/usr/lib/libcudnn_cnn_infer.so.8.6.0
/usr/lib/libcudnn_cnn_train.so.8.6.0
/usr/lib/libcudnn_ops_infer.so.8.6.0
/usr/lib/libcudnn_ops_train.so.8.6.0
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                    x86_64
CPU op-mode(s):                  32-bit, 64-bit
Address sizes:                   39 bits physical, 48 bits virtual
Byte Order:                      Little Endian
CPU(s):                          16
On-line CPU(s) list:             0-15
Vendor ID:                       GenuineIntel
Model name:                      11th Gen Intel(R) Core(TM) i9-11900 @ 2.50GHz
CPU family:                      6
Model:                           167
Thread(s) per core:              2
Core(s) per socket:              8
Socket(s):                       1
Stepping:                        1
CPU(s) scaling MHz:              72%
CPU max MHz:                     5200.0000
CPU min MHz:                     800.0000
BogoMIPS:                        4993.00
Flags:                           fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc art arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf tsc_known_freq pni pclmulqdq dtes64 monitor ds_cpl vmx smx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch cpuid_fault epb invpcid_single ssbd ibrs ibpb stibp ibrs_enhanced tpr_shadow vnmi flexpriority ept vpid ept_ad fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid mpx avx512f avx512dq rdseed adx smap avx512ifma clflushopt intel_pt avx512cd sha_ni avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves dtherm ida arat pln pts hwp hwp_notify hwp_act_window hwp_epp hwp_pkg_req avx512vbmi umip pku ospke avx512_vbmi2 gfni vaes vpclmulqdq avx512_vnni avx512_bitalg avx512_vpopcntdq rdpid fsrm md_clear flush_l1d arch_capabilities
Virtualization:                  VT-x
L1d cache:                       384 KiB (8 instances)
L1i cache:                       256 KiB (8 instances)
L2 cache:                        4 MiB (8 instances)
L3 cache:                        16 MiB (1 instance)
NUMA node(s):                    1
NUMA node0 CPU(s):               0-15
Vulnerability Itlb multihit:     Not affected
Vulnerability L1tf:              Not affected
Vulnerability Mds:               Not affected
Vulnerability Meltdown:          Not affected
Vulnerability Mmio stale data:   Mitigation; Clear CPU buffers; SMT vulnerable
Vulnerability Retbleed:          Mitigation; Enhanced IBRS
Vulnerability Spec store bypass: Mitigation; Speculative Store Bypass disabled via prctl
Vulnerability Spectre v1:        Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:        Mitigation; Enhanced IBRS, IBPB conditional, RSB filling, PBRSB-eIBRS SW sequence
Vulnerability Srbds:             Not affected
Vulnerability Tsx async abort:   Not affected

Versions of relevant libraries:
[pip3] numpy==1.24.2
[pip3] torch==2.0.0
[pip3] torchaudio==2.0.0
[pip3] torchvision==0.15.0
[pip3] triton==2.0.0
[conda] blas                      1.0                         mkl  
[conda] ffmpeg                    4.3                  hf484d3e_0    pytorch
[conda] mkl                       2021.4.0           h06a4308_640  
[conda] mkl-service               2.4.0           py310h7f8727e_0  
[conda] mkl_fft                   1.3.1           py310hd6ae3a3_0  
[conda] mkl_random                1.2.2           py310h00e6091_0  
[conda] numpy                     1.23.5          py310hd5efca6_0  
[conda] numpy-base                1.23.5          py310h8e6c178_0  
[conda] pytorch                   2.0.0           py3.10_cuda11.8_cudnn8.7.0_0    pytorch
[conda] pytorch-cuda              11.8                 h7e8668a_3    pytorch
[conda] pytorch-mutex             1.0                        cuda    pytorch
[conda] torchaudio                2.0.0               py310_cu118    pytorch
[conda] torchtriton               2.0.0                     py310    pytorch
[conda] torchvision               0.15.0              py310_cu118    pytorch
```

**Comment by FieldsMedal** [comment]
the following code can fix the  error.
![image](https://user-images.githubusercontent.com/28834157/229361641-5a9cb811-4f98-4723-b7a1-36a6b1699bea.png)

**Comment by ziadloo** [comment]
Thanks, @FieldsMedal . It does fix the issue but why? I mean what's wrong with instantiating the object right on the spot? Is this a Python issue?

**Comment by mthrok** [comment]
Hi

Thanks for the report and the solution.

Looks like it's an issue with the lifetime of Python object.
https://github.com/pybind/pybind11/discussions/4013

By doing `lm_weight=CustomLM()`, the `CustomLM` object is not referenced anywhere,
and it gets garbage-collected by the time its methods are called.

This seems to be a caveat generally applicable to any code with PyBind11.
It's been worked on https://github.com/pybind/pybind11/pull/2839.

I prepared a fix https://github.com/pytorch/audio/pull/3230

**Comment by mthrok** [comment]
Fixed by #3230
